File: sdk/typescript/src/hash.ts
import { sha256 } from "@noble/hashes/sha2.js";
import { bytesToHex } from "@noble/hashes/utils.js";

// Lowercase-hex SHA-256 of s.toLowerCase() — the lowercasing is applied to the
// whole string (not per field), matching Rocksky's server and every other SDK.
// Pure-JS digest (@noble/hashes) so this module stays browser-safe.
function sha256Lower(s: string): string {
  return bytesToHex(sha256(new TextEncoder().encode(s.toLowerCase())));
}

/** Identity hash of a song: sha256(lower("{title} - {artist} - {album}")). */
export function songHash(title: string, artist: string, album: string): string {
  return sha256Lower(`${title} - ${artist} - ${album}`);
}

/** Identity hash of an album: sha256(lower("{album} - {albumArtist}")). */
export function albumHash(album: string, albumArtist: string): string {
  return sha256Lower(`${album} - ${albumArtist}`);
}

/** Identity hash of an artist: sha256(lower(albumArtist)) — a single field. */
export function artistHash(albumArtist: string): string {
  return sha256Lower(albumArtist);
}

File: sdk/go/rocksky/hash.go
package rocksky

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
)

// sha256Lower returns the lowercase-hex SHA-256 of s.ToLower() — the lowercasing
// is applied to the whole string (not per field), matching Rocksky's server and
// every other Rocksky SDK.
func sha256Lower(s string) string {
	sum := sha256.Sum256([]byte(strings.ToLower(s)))
	return hex.EncodeToString(sum[:])
}

// SongHash is a song/track's identity: sha256(lower("{title} - {artist} - {album}")).
// Uses the per-track artist, not the album artist.
func SongHash(title, artist, album string) string {
	return sha256Lower(title + " - " + artist + " - " + album)
}

// AlbumHash is an album's identity: sha256(lower("{album} - {albumArtist}")).
func AlbumHash(album, albumArtist string) string {
	return sha256Lower(album + " - " + albumArtist)
}

// ArtistHash is an artist's identity: sha256(lower(albumArtist)) — a single field.
func ArtistHash(albumArtist string) string {
	return sha256Lower(albumArtist)
}
